openeo-geotrellis/src/main/scala/org/openeo/geotrellis/ComputeStatsGeotrellisAdapter.scala

```diff
@@ -29,6 +29,108 @@ import scala.collection.JavaConverters._
+  }
+
+  def parsePolygonWkt(polygonWkt: String): MultiPolygon = {
+    val geometry: Geometry = polygonWkt.parseWKT()
+    geometry match {
+      case multiPolygon: MultiPolygon => multiPolygon
+      case _ => MultiPolygon(geometry.asInstanceOf[Polygon])
+    }
+  }
+
+  def fromVectorFile(vector_file: String): ProjectedPolygons = {
+    val vectorUrl = try {
+      new URL(vector_file)
+    } catch {
+      case _: MalformedURLException => new URL(s"file://$vector_file")
+    }
+
+    val filename = vectorUrl.getPath.split("/").last
+
+    if (filename.endsWith(".shp")) readSimpleFeatures(vectorUrl)
+    else readMultiPolygonsFromGeoJson(vectorUrl)
+  }
+
+  // adapted from Geotrellis' ShapeFileReader to avoid having too much in memory
+  private def readSimpleFeatures(shpUrl: URL): ProjectedPolygons = {
+    val ds = new ShapefileDataStore(shpUrl)
+    val ftItr: SimpleFeatureIterator = ds.getFeatureSource.getFeatures.features
+
+    try {
+      val featureCount = ds.getCount(Query.ALL)
+      require(featureCount < Int.MaxValue)
+
+      val simpleFeatures = new Array[MultiPolygon](featureCount.toInt)
+
+      for (i <- simpleFeatures.indices) {
+        val multiPolygon = ftItr.next().getAttribute(0) match {
+          case multiPolygon: MultiPolygon => multiPolygon
+          case polygon: Polygon => MultiPolygon(polygon)
+        }
+
+        simpleFeatures(i) = multiPolygon
+      }
+
+      // FIXME: read it from the shp and default to LatLng
+      ProjectedPolygons(simpleFeatures, LatLng)
+    } finally {
+      ftItr.close()
+      ds.dispose()
+    }
+  }
+
+  private def readMultiPolygonsFromGeoJson(geoJsonUrl: URL): ProjectedPolygons = {
+    // FIXME: stream it instead
+    val src = Source.fromURL(geoJsonUrl)
+
+    val multiPolygons = try {
+      val geoJson = src.mkString
+
+      def children(geometryCollection: GeometryCollection): Stream[Geometry] = {
+        def from(i: Int): Stream[Geometry] =
+          if (i >= geometryCollection.getNumGeometries) Stream.empty
+          else geometryCollection.getGeometryN(i) #:: from(i + 1)
+
+        from(0)
+      }
+
+      def asMultiPolygons(geometry: Geometry): Array[MultiPolygon] = geometry match {
+        case polygon: Polygon => Array(MultiPolygon(polygon))
+        case multiPolygon: MultiPolygon => Array(multiPolygon)
+        case geometryCollection: GeometryCollection => children(geometryCollection).map {
+          case polygon: Polygon => MultiPolygon(polygon)
+          case multiPolygon: MultiPolygon => multiPolygon
+        }.toArray
+      }
+
+      try {
+        asMultiPolygons(geoJson.parseGeoJson[Geometry]())
+      } catch {
+        case _: DecodingFailure =>
+          val featureCollection = geoJson.parseGeoJson[JsonFeatureCollection]()
+          featureCollection.getAllGeometries()
+            .flatMap(asMultiPolygons)
+            .toArray
+      }
+    } finally src.close()
+
+    ProjectedPolygons(multiPolygons, LatLng)
+  }
+
+}
+
```

Comment:
In Scala, unlike Java, you can have multiple public classes within a single file but `ProjectedPolygons` is large enough to warrant a separate file; put `ProjectedPolygons` and its companion object into a file called ProjectedPolygons.scala next to this file.

Reply:
indeed, `ProjectedPolygons` turned out larger than I anticipated, 
will move to new file